library/alloc/src/macros.rs

```diff
@@ -101,9 +101,18 @@ macro_rules! vec {
 /// ```
 #[macro_export]
 #[stable(feature = "rust1", since = "1.0.0")]
+#[allow_internal_unstable(fmt_as_str)]
 macro_rules! format {
     ($($arg:tt)*) => {{
-        let res = $crate::fmt::format($crate::__export::format_args!($($arg)*));
-        res
+        let r = match $crate::__export::format_args!($($arg)*) {
+            args => {
+                if let Some(s) = args.as_str() {
+                    $crate::borrow::ToOwned::to_owned(s)
+                } else {
+                    $crate::fmt::format(args)
+                }
```

Comment:
Maybe this branch should go in `fmt::format` instead, so everything gets the benefit and not just `format!`?

Reply:
`fmt::format` is not `#[inline]`, so it can't be inlined cross-crate.

Reply:
If we have [inline const expressions](https://github.com/rust-lang/rfcs/pull/2920), we could try that.
Because `fmt::Arguments` is only initilized at compile time by `format_args!` only, this branch should not fall to be compiled.

Reply:
That seems like a good thing, though - this way you have less code in the instruction cache in case this is in a branch that's not taken.

Reply:
Would this branch appear in runtime?

Reply:
The `Arguments::as_str` is a const fn, it shouldn't appear at runtime.
If you could come up with an example, I could add codegen for it.

Reply:
Inline const expressions can already be emulated:
```rust
const MYCONST: Type = expr;
MYCONST
```
So, "If we have inline const expressions, we could try that" implies that we can already try it now.

Reply:
> The Arguments::as_str is a const fn, it shouldn't appear at runtime.

Whether or not a function is `const` has no effect at all for codegen and runtime behavior.

All `const fn` changes compared to `fn` is that now you are allowed to call this function in a `const`/`static` initializer. That's it. I don't understand why people keep thinking that it would do any more than that...

Reply:
Nice trick about emulating inline const expr! I'll try that.

> you are allowed to call this function in a const/static initializer.
Maybe I misunderstand. But given `const fn foo(usize) -> usize`, does it guarantee `foo(42)` a constant at compile time ?

Reply:
No, while rustc may do that as an optimizations in some cases in the future, it currently doesn't. Instead you have to hope that LLVM performs the inlining + const folding.

Reply:
> All const fn changes compared to fn is that now you are allowed to call this function in a const/static initializer. That's it. I don't understand why people keep thinking that it would do any more than that...

Oh, I always thought that `const` means all of these happens in compile time and none of it goes to runtime except those called in non-const places. I thought it could even let rustc knows that there are no side-effects at runtime for the const function.

Reply:
> But what's the point of a const function then?

It makes code like this compile:
```rust
const FOO: i32 = my_const_fn(13);
```
That code is rejected unless `my_const_fn` is a `const fn`.

That is the *only* effect `const` has on `fn`.

>  I always thought that const means all of these happens in compile time and none of it goes to runtime except those called in non-const places.

Code in const "places" (usually called const contexts( *must* run at compiletime. So this statement is true but I don't think it means what you think it means. In particular, `format` is impossible to use in const contexts:
```rust
const FOO: String = format!("abc"); // does not compile
```

Reply:
@RalfJung Thanks for clarifying. I think we could do a poll on this. I bet most people think `const fn` will allow optimizations like these, I bet @lzutao had the same thing in mind too. Haha, would be interesting to talk about `const fn` in rustconf, @RalfJung maybe you want to bust this myth there?

Reply:
Sometimes the majority is wrong. ;)

Next RustConf is in a year.^^ I might just write a blog post about it, though that is not really a sustainable solution... and I have 3 or 4 other blog posts that I want to write already in the queue.

Reply:
I wonder if we could ask this question the next time we sent out our surveys. Who usually do the survey and why not we put this as a question? I bet the majority may still be wrong next year. Me trying to get your words into quote of the week.

Reply:
I'll mark it as resolved since I answered it. If anyone want to bring it up, please create an new review.